From 81435a7eacc7116345ffdef9b62d939f8997cc8e Mon Sep 17 00:00:00 2001
From: Yi-Chia Hsieh <Yi-Chia.Hsieh@mediatek.com>
Date: Fri, 15 Aug 2025 17:38:13 -0700
Subject: [PATCH 116/128] mtk: mac80211: modify struct for broadcast TWT

Restructure ieee80211_twt_params. The orignal structure only consider
the case of individual TWT and they do not fit for broadcast TWT.

Signed-off-by: Yi-Chia Hsieh <Yi-Chia.Hsieh@mediatek.com>
---
 include/linux/ieee80211-he.h | 18 ++++++++++++++----
 net/mac80211/trace.h         |  8 ++++----
 2 files changed, 18 insertions(+), 8 deletions(-)

diff --git a/include/linux/ieee80211-he.h b/include/linux/ieee80211-he.h
index a08c446f..b652ede7 100644
--- a/include/linux/ieee80211-he.h
+++ b/include/linux/ieee80211-he.h
@@ -46,10 +46,20 @@ enum ieee80211_twt_setup_cmd {
 
 struct ieee80211_twt_params {
 	__le16 req_type;
-	__le64 twt;
-	u8 min_twt_dur;
-	__le16 mantissa;
-	u8 channel;
+	union {
+		struct {
+			__le64 twt;
+			u8 min_twt_dur;
+			__le16 mantissa;
+			u8 channel;
+		} __packed indiv;
+		struct {
+			__le16 twt;
+			u8 min_twt_dur;
+			__le16 mantissa;
+			__le16 btwt_info;
+		} __packed bcast;
+	};
 } __packed;
 
 struct ieee80211_twt_setup {
diff --git a/net/mac80211/trace.h b/net/mac80211/trace.h
index e219a2a1..1373cb0d 100644
--- a/net/mac80211/trace.h
+++ b/net/mac80211/trace.h
@@ -2511,10 +2511,10 @@ TRACE_EVENT(drv_add_twt_setup,
 		__entry->dialog_token = twt->dialog_token;
 		__entry->control = twt->control;
 		__entry->req_type = twt_agrt->req_type;
-		__entry->twt = twt_agrt->twt;
-		__entry->duration = twt_agrt->min_twt_dur;
-		__entry->mantissa = twt_agrt->mantissa;
-		__entry->channel = twt_agrt->channel;
+		__entry->twt = twt_agrt->indiv.twt;
+		__entry->duration = twt_agrt->indiv.min_twt_dur;
+		__entry->mantissa = twt_agrt->indiv.mantissa;
+		__entry->channel = twt_agrt->indiv.channel;
 	),
 
 	TP_printk(
-- 
2.45.2

